### PR TITLE
Uncle inclusion in block authoring.

### PR DESCRIPTION
**Pull request by gavofyork**

### CODE REVIEW

File: ethcore/src/client.rs [hunk 1]
```diff
@@ -203,7 +203,7 @@ pub struct Client {
 	extra_data: RwLock<Bytes>,
 }
 
-const HISTORY: u64 = 1000;
```
Comment (arkpar) [inline ethcore/src/client.rs]: Unintended?

**Comment by arkpar** [comment]
Needs merge

File: ethcore/src/block.rs [hunk 2]
```diff
@@ -220,8 +220,8 @@ impl<'x> OpenBlock<'x> {
 	/// NOTE Will check chain constraints and the uncle number but will NOT check
 	/// that the header itself is actually valid.
 	pub fn push_uncle(&mut self, valid_uncle_header: Header) -> Result<(), BlockError> {
-		if self.block.base.uncles.len() >= self.engine.maximum_uncle_count() {
-			return Err(BlockError::TooManyUncles(OutOfBounds{min: None, max: Some(self.engine.maximum_uncle_count()), found: self.block.base.uncles.len()}));
+		if self.block.base.uncles.len() + 1 > self.engine.maximum_uncle_count() {
```
Comment (gavofyork) [inline ethcore/src/block.rs]: Make this a little more readable

File: ethcore/src/blockchain/blockchain.rs [hunk 3]
```diff
@@ -227,6 +227,24 @@ impl BlockProvider for BlockChain {
 
 const COLLECTION_QUEUE_SIZE: usize = 8;
 
+pub struct AncestryIter<'a> {
```
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: Iterator for traversing ancestry.

File: ethcore/src/client.rs [hunk 4]
```diff
@@ -200,6 +201,7 @@ pub struct Client<V = CanonVerifier> where V: Verifier {
 	panic_handler: Arc<PanicHandler>,
 
 	// for sealing...
+	sealing_enabled: AtomicBool,
```
Comment (gavofyork) [inline ethcore/src/client.rs]: Add flag for remembering if we should be sealing or not.

File: ethcore/src/client.rs [hunk 5]
```diff
@@ -398,7 +401,7 @@ impl<V> Client<V> where V: Verifier {
 			}
 		}
 
-		if self.chain_info().best_block_hash != original_best {
+		if self.chain_info().best_block_hash != original_best && self.sealing_enabled.load(atomic::Ordering::Relaxed) {
```
Comment (gavofyork) [inline ethcore/src/client.rs]: Avoid calling expensive `prepare_sealing` if we haven't already been asked for a work package from the miner.

File: ethcore/src/client.rs [hunk 6]
```diff
@@ -493,6 +496,8 @@ impl<V> Client<V> where V: Verifier {
 	/// Grab the `ClosedBlock` that we want to be sealed. Comes as a mutex that you have to lock.
 	pub fn sealing_block(&self) -> &Mutex<Option<ClosedBlock>> {
 		if self.sealing_block.lock().unwrap().is_none() {
+			self.sealing_enabled.store(true, atomic::Ordering::Relaxed);
```
Comment (gavofyork) [inline ethcore/src/client.rs]: We've been asked for a work package. Start automatically sealing each block.

File: ethcore/src/engine.rs [hunk 7]
```diff
@@ -47,6 +47,8 @@ pub trait Engine : Sync + Send {
 	fn maximum_extra_data_size(&self) -> usize { decode(&self.spec().engine_params.get("maximumExtraDataSize").unwrap()) }
 	/// Maximum number of uncles a block is allowed to declare.
 	fn maximum_uncle_count(&self) -> usize { 2 }
+	/// The number of generations back that uncles can be.
+	fn maximum_uncle_age(&self) -> usize { 6 }
```
Comment (gavofyork) [inline ethcore/src/engine.rs]: Additional chain parameter

File: ethcore/src/header.rs [hunk 8]
```diff
@@ -70,6 +70,25 @@ pub struct Header {
 	pub bare_hash: RefCell<Option<H256>>,
 }
 
+impl PartialEq for Header {
+	fn eq(&self, c: &Header) -> bool {
```
Comment (gavofyork) [inline ethcore/src/header.rs]: comparison *without* the memoisation fields.

**Pull request merged**